src/algebra/modeq.lean

```diff
@@ -0,0 +1,197 @@
+
+## See also
+
+`smodeq` is a generalisation to arbitrary submodules.
+
+## TODO
+
+Delete `int.modeq` in favour of `add_comm_group.modeq`. Generalise `smodeq` to `add_subgroup` and
+redefine `add_comm_group.modeq` using it. Once this is done, we can rename `add_comm_group.modeq`
+to `add_subgroup.modeq` and multiplicativise it.
+-/
+
+namespace add_comm_group
+variables {α : Type*}
+
+section add_comm_group
+variables [add_comm_group α] {p a a₁ a₂ b b₁ b₂ c : α} {n : ℤ}
+
+/-- `a ≡ b [PMOD p]` means that `b` is congruent to `a` modulo `p`.
+
+Equivalently (as shown in `algebra.order.to_interval_mod`), `b` does not lie in the open interval
+`(a, a + p)` modulo `p`, or `to_Ico_mod hp a` disagrees with `to_Ioc_mod hp a` at `b`, or
+`to_Ico_div hp a` disagrees with `to_Ioc_div hp a` at `b`. -/
+def modeq (p a b : α) : Prop := ∃ n : ℤ, b - a = n • p
+
+notation a ` ≡ `:50 b ` [PMOD `:50 p `]`:0 := modeq p a b
+
+@[refl, simp] lemma modeq_refl (a : α) : a ≡ a [PMOD p] := ⟨0, by simp⟩
+
+lemma modeq_rfl : a ≡ a [PMOD p] := modeq_refl _
+
+lemma modeq_comm : a ≡ b [PMOD p] ↔ b ≡ a [PMOD p] :=
+(equiv.neg _).exists_congr_left.trans $ by simp [modeq, ←neg_eq_iff_eq_neg]
+
+alias modeq_comm ↔ modeq.symm _
+
+attribute [symm] modeq.symm
+
+@[trans] lemma modeq.trans : a ≡ b [PMOD p] → b ≡ c [PMOD p] → a ≡ c [PMOD p] :=
+λ ⟨m, hm⟩ ⟨n, hn⟩, ⟨m + n, by simp [add_smul, ←hm, ←hn]⟩
+
+instance : is_refl _ (modeq p) := ⟨modeq_refl⟩
+
+@[simp] lemma neg_modeq_neg : -a ≡ -b [PMOD p] ↔ a ≡ b [PMOD p] :=
+modeq_comm.trans $ by simp [modeq]
+
+alias neg_modeq_neg ↔ modeq.of_neg modeq.neg
+
+@[simp] lemma modeq_neg : a ≡ b [PMOD -p] ↔ a ≡ b [PMOD p] :=
+modeq_comm.trans $ by simp [modeq, ←neg_eq_iff_eq_neg]
+
+alias modeq_neg ↔ modeq.of_neg' modeq.neg'
+
+@[simp] lemma zsmul_modeq_zsmul [no_zero_smul_divisors ℤ α] (hn : n ≠ 0) :
+  n • a ≡ n • b [PMOD (n • p)] ↔ a ≡ b [PMOD p] :=
+exists_congr $ λ m, by rw [←smul_sub, smul_comm, smul_right_inj hn]; apply_instance
+
+alias zsmul_modeq_zsmul ↔ modeq.zsmul_cancel _
+
+lemma modeq_sub (a b : α) : a ≡ b [PMOD b - a] := ⟨1, (one_smul _ _).symm⟩
+
+@[simp] lemma modeq_zero : a ≡ b [PMOD 0] ↔ a = b := by simp [modeq, sub_eq_zero, eq_comm]
+
+@[simp] lemma zsmul_modeq_zero (n : ℤ) : n • p ≡ 0 [PMOD p] := ⟨-n, by simp⟩
+@[simp] lemma self_modeq_zero : p ≡ 0 [PMOD p] := ⟨-1, by simp⟩
+
+lemma add_zsmul_modeq : a + n • p ≡ a [PMOD p] := ⟨-n, by simp⟩
+lemma zsmul_add_modeq : n • p + a ≡ a [PMOD p] := ⟨-n, by simp⟩
+
+namespace modeq
+
+protected lemma of_zsmul : a ≡ b [PMOD (n • p)] → a ≡ b [PMOD p] :=
+λ ⟨m, hm⟩, ⟨m * n, by rwa [mul_smul]⟩
+
+protected lemma zsmul : a ≡ b [PMOD p] → n • a ≡ n • b [PMOD (n • p)] :=
+Exists.imp $ λ m hm, by rw [←smul_sub, hm, smul_comm]
+
+@[simp] protected lemma add_iff_left :
+  a₁ ≡ b₁ [PMOD p] → (a₁ + a₂ ≡ b₁ + b₂ [PMOD p] ↔ a₂ ≡ b₂ [PMOD p]) :=
+λ ⟨m, hm⟩, (equiv.add_left m).symm.exists_congr_left.trans $
+  by simpa [add_sub_add_comm, hm, add_smul]
```

Comment:
Note to myself: these don't hold for a non-normal subgroup in a (nonabelian) group, but still hold for a normal subgroup that is not necessarily `zmultiples`.